Refactorings in preparation for parallel visualization execution

hubertp opened the pull request:
### Pull Request Description

[Work on reactive observers for visualizations](https://github.com/enso-org/enso/pull/13219) revealed a number of blockers once one attempts to execute visualizations in parallel. This PR extracts the less controversial parts that became obvious once we opened a pandora's box, as well as a number of startup improvements:

- push context request does not trigger execution by default (it typically lacks `IdMap` necessary for instrumentation at that stage)
- split context lock to read/write one to reduce lock contention and ensure correctness
- changing methods on `ExecutionService` to not block
- make `HostClassLoader` thread-safe
- reduce write compilation lock contention by first attempting to evaluate an expression with a read lock only
- fixed a bug that potentially invalidates visualizations' state too often

### Important Notes

Should be easier to review than https://github.com/enso-org/enso/pull/13219 that went through a number of experiments.
_Some_ startup improvements should be visible from this change already as we reduce lock contention.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.

hubertp committed:
Minor refactorings

hubertp committed:
Fix bug in sync expressions/visualizations

hubertp committed:
Modify ExecutionService API to return futures

Most of `ExecutionService` now returns Futures that can be easier to
compose and execute via threadpools.

hubertp committed:
Optional execution in PushContext request

By default, no execution is triggered by push context request.

hubertp committed:
fix NI build

hubertp committed:
drop unused parameter

hubertp committed:
fix typo

hubertp committed:
Infer diagnostics in Truffle context

When diagnostics are being gathered outside of ThreadManager, execution
will blow up due to Truffle restrictions.

hubertp committed:
Split context lock into read/write one

Reduce lock contention by splitting context lock into a read/write one.
Context lock has been slowly ignored in various jobs in favour of
improved performance. It shouldn't be, as most of the time we only need
a read context lock.

hubertp committed:
nit

hubertp committed:
Make HostClassLoader thread-safe

hubertp committed:
nit

hubertp committed:
compile tests

hubertp committed:
Typo: run and try to set state

hubertp committed:
Recompute should be under a proper context lock

hubertp committed:
Adapt tests

hubertp committed:
Fix typoe

hubertp committed:
fix tests

hubertp committed:
Export Widgets in Visualization's Main module

Should reduce lock contention on startup.

hubertp committed:
remove ExecutionService.resultOf

hubertp committed:
s/CompletableFuture/CompletionStage

hubertp committed:
Merge branch 'develop' into wip/hubert/10525-prep-parallel-exec

GregoryTravis reviewed (approved):
Approval for .enso changes

hubertp committed:
adapt tests

hubertp committed:
Merge branch 'develop' into wip/hubert/10525-prep-parallel-exec

hubertp committed:
merge conflict leftover

hubertp committed:
Fix remaining ExecutionException leaks

When reporting error diagnostics we need to analyze the error underlying
the `ExecutionException`, not the exception itself.

hubertp committed:
Bump NI max size on Mac

It now says 427, so adding +30 as a general rule.
Dunno on why would increase be needed.

JaroslavTulach commented on engine/runtime-instrument-common/src/main/java/org/enso/interpreter/instrument/RuntimeCache.java:
```diff
@@ -31,16 +31,13 @@ public final class RuntimeCache implements java.util.function.Function<String, O
-      expressions.put(key, ref);
-      return false;
     }
+    return false;
```
Nit: the previous version with two `return` statement in each `if` branch felt more functional to me.

JaroslavTulach commented on engine/runtime-instrument-common/src/main/java/org/enso/interpreter/service/ExecutionService.java:
```diff
@@ -112,21 +118,25 @@ public EnsoContext getContext() {
   }
 
-  public FunctionCallInstrumentationNode.FunctionCall prepareFunctionCall(
+  public CompletionStage<FunctionCallInstrumentationNode.FunctionCall> prepareFunctionCall(
```
Glad to see `CompletionStage` working.

JaroslavTulach commented on engine/runtime-instrument-common/src/main/java/org/enso/interpreter/service/ExecutionService.java:
```diff
@@ -590,20 +634,10 @@ static <E extends Throwable> E raise(Class<E> type, Throwable ex) throws E {
     return context.getThreadManager().submit(c);
   }
 
-  private static <T> T resultOf(Future<T> future) {
```
Removal of `resultOf` blocking call marks `ExecutionService` as _fully reactive_ (e.g. non-blocking) API. Good.

JaroslavTulach commented on engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/WarningPreview.scala:
```diff
@@ -16,19 +16,27 @@ object WarningPreview {
+        )
+      )
+
+    visualizationResultFuture.thenApply(visualizationResult => {
```
It's reactive!

JaroslavTulach commented on engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/execution/ReentrantLocking.scala:
```diff
@@ -214,9 +216,44 @@ class ReentrantLocking extends Locking {
+  ): T = {
+    val contextLock                = lock.asInstanceOf[ContextLockImpl]
+    var contextLockTimestamp: Long = 0
+    val writeLock                  = contextLock.lock.writeLock()
```
- I don't see any [assertNotLocked](https://github.com/enso-org/enso/blob/6edcdaf8dce7d7b7a5f7cbd792c817f51e1d6ccd/engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/execution/ReentrantLocking.scala#L386)
- clearly asking for a `withWriteContextLock` while already holding `withReadContextLock` would be _deadlock prone behavior_
- I'd assert against that

hubertp replied:
`ReentrantReadWriteLock` does not allow for upgrading locks as per [javadoc](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/locks/ReentrantReadWriteLock.html) so it felt unnecessary.

JaroslavTulach replied:
OK

JaroslavTulach commented on engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/job/ProgramExecutionSupport.scala:
```diff
@@ -168,6 +160,7 @@ object ProgramExecutionSupport {
           onCachedValueCallback,
           onExecutedVisualizationCallback
         )
+        pending.toCompletableFuture.get()
```
- Blocking, eh!
- Is blocking only a temporary (to minimize the size of the PR)
- Or is the blocking in `ProgramExecutionSupport` to stay?

hubertp replied:
I didn't need it. If you feel strongly we can propagate it further or leave it for later.

JaroslavTulach replied:
Certainly: leave for later (or leave it forever as it is).

JaroslavTulach commented on engine/runtime-instrument-common/src/main/java/org/enso/interpreter/service/ExecutionService.java:
```diff
@@ -552,12 +597,11 @@ public boolean isExitException(AbstractTruffleException ex) {
-  public String getExceptionMessage(AbstractTruffleException panic) {
-    var future = submitExecution(() -> computeExceptionMessage(panic));
-    return resultOf(future);
+  public CompletionStage<String> getExceptionMessage(AbstractTruffleException panic) {
```
This one would very likely work better as:
```java
CompletionStage<String> getExceptionMessage(CompletionStage<AbstractTruffleException> panic)
```
or something a like that being _part of the reactive flow_. It makes little sense to `ex = toCompletableFuture().get()` and then enter again the _guest code_ thread by `getExceptionMessage(ex)`.

The same comment applies to `typeOfValue` and other methods that inspect _result of guest code_ execution.

JaroslavTulach reviewed (approved):
- I haven't tried to look at language server side of the change much
- `ExecutionService` changes towards `CompletionStage` seem good
- possibly they could be even more _reactive_ to avoid on/off/on switching

hubertp committed:
Merge branch 'develop' into wip/hubert/10525-prep-parallel-exec

hubertp committed:
Merge branch 'develop' into wip/hubert/10525-prep-parallel-exec

hubertp commented:
I will address the remaining comments in follow up PRs.
